### PR TITLE
Add rollout workflow for s3

### PR DESCRIPTION
**Pull request by daanpersoons-appwise**

### CODE REVIEW

File: .github/workflows/node-publish-docs.yml [hunk 1]
```diff
@@ -10,7 +10,7 @@ on:  # yamllint disable-line rule:truthy
     inputs:
       dist-path:
         description: Path to build output
-        default: 'packages/docs/.vitepress/dist/'
```
Comment (djbe) [inline .github/workflows/node-publish-docs.yml:13]: Move this to a separate PR, it doesn't belong here.

File: .github/workflows/s3-workflow-rollout-to-scaleway.yml [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+          access_key: ${{ secrets.SCW_ACCESS_KEY_ID }}
+          secret_key: ${{ secrets.SCW_SECRET_ACCESS_KEY }}
+
+      - name: Check existing/new buckets and apply lifecycle and cors
+        run: |
+          for e in s3/${{ matrix.deployment }}/*;
+          do
+            s3cmd mb s3://s3-$(basename -- $e)-${{ matrix.deployment }} || true
+            s3cmd setcors $e/cors.xml s3://s3-$(basename -- $e)-${{ matrix.deployment }};
+            s3cmd setlifecycle $e/lifecycle.xml s3://s3-$(basename -- $e)-${{ matrix.deployment }};
```
Comment (djbe) [inline .github/workflows/s3-workflow-rollout-to-scaleway.yml:68]: I would add an if test around here to check if the lifecycle file exists. If not, default to one we embed in the action itself (see the linter workflow), that's simply the delete temp uploads one.

That way we don't have to commit 999x copies of the same lifecycle file, and are still able to have a custom one for very specific buckets.

File: .github/workflows/s3-workflow-rollout-to-scaleway.yml [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        deployment: ${{ fromJson(needs.generate-matrix.outputs.buckets) }}
+    if: ${{ needs.generate-matrix.outputs.buckets != '[]' }}]
+    environment: ${{ matrix.deployment }}
+    steps:
+      - name: Checkout s3 folder
+        uses: actions/checkout@v4
+        with:
+          sparse-checkout: |
+            s3
+
```
Comment (djbe) [inline .github/workflows/s3-workflow-rollout-to-scaleway.yml:53]: Nitpick: remove the newlines between "steps", make it obvious what belongs together.

File: .github/workflows/s3-workflow-rollout-to-scaleway.yml [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+      - name: Generate Matrix
+        id: filtered
+        uses: dorny/paths-filter@v3
+        with:
+          filters: .github/s3-rollout-filters.yaml
+
+  rollout-buckets:
+    runs-on: ubuntu-latest
+    needs: [generate-matrix]
+    name: S3CMD
```
Comment (djbe) [inline .github/workflows/s3-workflow-rollout-to-scaleway.yml:41]: Change this to "Rollout {{deployment}} buckets` so we can easily see what's running.

File: .github/workflows/s3-workflow-rollout-to-scaleway.yml [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+      - name: Set up S3cmd cli tool for environment
+        uses: s3-actions/s3cmd@main
+        with:
+          provider: scaleway
+          region: ${{ inputs.scaleway-region }}
+          access_key: ${{ secrets.SCW_ACCESS_KEY_ID }}
+          secret_key: ${{ secrets.SCW_SECRET_ACCESS_KEY }}
+
+      - name: Check existing/new buckets and apply lifecycle and cors
```
Comment (djbe) [inline .github/workflows/s3-workflow-rollout-to-scaleway.yml:62]: Move this code into a separate rollout s3 job, which could be reused for other providers. Same as we did with docker, k8s, etc…

**Review by djbe — CHANGES_REQUESTED** [review]

File: .github/workflows/s3-workflow-rollout-to-scaleway.yml (hunk 3 above)
Comment (daanpersoons-appwise) [reply .github/workflows/s3-workflow-rollout-to-scaleway.yml:53]: @djbe done, is there a reason for this or is it just code aesthetics?

File: .github/workflows/s3-workflow-rollout-to-scaleway.yml (hunk 4 above)
Comment (daanpersoons-appwise) [reply .github/workflows/s3-workflow-rollout-to-scaleway.yml:41]: You mean plain text {{deployment}} or the variable?
Variable is impossible because it gets generated in that job

File: .github/workflows/s3-workflow-rollout-to-scaleway.yml (hunk 3 above)
Comment (djbe) [reply .github/workflows/s3-workflow-rollout-to-scaleway.yml:53]: Mostly readability, what I try to have in code/configs is clear separation between "logical blocks". It's a bit dependent on the situation, I don't think there's a clear "rule". In this case we have the start/triggers, then the setup with the matrix, and then finally the workflow itself.

File: .github/workflows/s3-workflow-rollout-to-scaleway.yml (hunk 4 above)
Comment (djbe) [reply .github/workflows/s3-workflow-rollout-to-scaleway.yml:41]: The variable. It should't be impossible? Yes it gets generated in the previous job, but we can use it in THIS job (which gets repeated for each deployment).

File: .github/workflows/workflow-rollout-s3-digitalocean.yml [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+
+#
+# Build the docker image, and deploy it to the given environment.
+#
+
+on:  # yamllint disable-line rule:truthy
+  workflow_call:
+    inputs:
+      # digitalocean information
+      do-region:
```
Comment (djbe) [inline .github/workflows/workflow-rollout-s3-digitalocean.yml:12]:
```suggestion
      digitalocean-region:
```

File: .github/workflows/workflow-rollout-s3-digitalocean.yml
Comment (djbe) [inline .github/workflows/workflow-rollout-s3-digitalocean.yml:1]: I think we should name these `s3-workflow-rollout-to-…`

File: .github/workflows/s3-rollout.yml [hunk 7]
```diff
@@ -0,0 +1,90 @@
+---
+name: Rollout S3 buckets
+on:  # yamllint disable-line rule:truthy
+  workflow_call:
+    inputs:
+      # Scaleway information
+      s3-region:
```
Comment (djbe) [inline .github/workflows/s3-rollout.yml:7]: No need to prefix these with `s3`, we don't do it for the other (internal) workflows. This is a workflow for s3 stuff, so it's obvious from that.
```suggestion
      region:
```

File: .github/workflows/s3-rollout.yml [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+        description: Region to deploy to
+        required: true
+        type: string
+      s3-provider:
+        description: Provider to deploy to
+        required: true
+        type: string
+    secrets:
+      S3_ACCESS_KEY:
```
Comment (djbe) [inline .github/workflows/s3-rollout.yml:16]: Same:
```suggestion
      ACCESS_KEY:
```

File: .github/workflows/s3-rollout.yml [hunk 9]
```diff
@@ -0,0 +1,90 @@
+---
+name: Rollout S3 buckets
+on:  # yamllint disable-line rule:truthy
+  workflow_call:
+    inputs:
+      # Scaleway information
```
Comment (djbe) [inline .github/workflows/s3-rollout.yml:6]:
```suggestion
      # Provider information
```

File: .github/workflows/s3-rollout.yml [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+  generate-matrix:
+    name: Generate job matrices
+    runs-on: ubuntu-latest
+    outputs:
+      buckets: ${{ steps.filtered.outputs.changes }}
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v4
+        with:
+          sparse-checkout: |
```
Comment (djbe) [inline .github/workflows/s3-rollout.yml:48]: Good idea, we may want to also use this for the k8s workflows (in a different PR)

File: .github/workflows/workflow-rollout-s3-scaleway.yml [hunk 11]
```diff
@@ -0,0 +1,32 @@
+---
+name: Build & deploy to environment (Scaleway)
```
Comment (djbe) [inline .github/workflows/workflow-rollout-s3-scaleway.yml:2]:
```suggestion
name: Rollout s3 object storage (Scaleway)
```

File: .github/workflows/workflow-rollout-s3-scaleway.yml [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+
+#
+# Build the docker image, and deploy it to the given environment.
```
Comment (djbe) [inline .github/workflows/workflow-rollout-s3-scaleway.yml:5]:
```suggestion
# Create & configure s3 buckets for each environment.
```

File: .github/workflows/workflow-rollout-s3-digitalocean.yml [hunk 13]
```diff
@@ -0,0 +1,32 @@
+---
+name: Build & deploy to environment (Digitalocean)
+
+#
+# Build the docker image, and deploy it to the given environment.
```
Comment (djbe) [inline .github/workflows/workflow-rollout-s3-digitalocean.yml:5]:
```suggestion
# Create & configure s3 buckets for each environment.
```

File: .github/workflows/workflow-rollout-s3-digitalocean.yml [hunk 14]
```diff
@@ -0,0 +1,32 @@
+---
+name: Build & deploy to environment (Digitalocean)
```
Comment (djbe) [inline .github/workflows/workflow-rollout-s3-digitalocean.yml:2]:
```suggestion
name: Rollout s3 object storage (DigitalOcean)
```

File: .github/workflows/s3-rollout.yml [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+      - name: Generate Matrix
+        id: filtered
+        uses: dorny/paths-filter@v3
+        with:
+          filters: .github/s3-rollout-filters.yaml
+
+  rollout-buckets:
+    runs-on: ubuntu-latest
+    needs: [generate-matrix]
+    name: S3CMD
```
Comment (djbe) [inline .github/workflows/s3-rollout.yml:60]:
```suggestion
    name: Rollout s3 for ${{ matrix.deployment }}
```

File: .github/workflows/s3-rollout.yml [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+        with:
+          provider: ${{ inputs.s3-provider }}
+          region: ${{ inputs.s3-region }}
+          access_key: ${{ secrets.S3_ACCESS_KEY}}
+          secret_key: ${{ secrets.S3_SECRET_KEY }}
+      - name: Check existing/new buckets and apply lifecycle and cors
+        run: |
+          for e in s3/${{ matrix.deployment }}/*;
+          do
+            s3cmd mb s3://s3-$(basename -- $e)-${{ matrix.deployment }} || true
```
Comment (djbe) [inline .github/workflows/s3-rollout.yml:83]: You may want to "calculate" the bucket name and store it in a variable, you're using it multiple times. So
```suggestion
            BUCKET_NAME="s3-$(basename -- $e)-${{ matrix.deployment }}"
            s3cmd mb s3://$BUCKET_NAME || true
```

**Review by djbe — CHANGES_REQUESTED** [review]

File: .github/workflows/s3-rollout.yml (hunk 7 above)
Comment (daanpersoons-appwise) [reply .github/workflows/s3-rollout.yml:7]: Same for provider then?

**Comment by djbe** [comment]
- I've refactored out the apply step into a dedicated action (like I've done with the k8s steps), you can find it here: https://github.com/wisemen-digital/devops-ga-s3-apply
- Reworked to use the new action & the custom filter action. This has changed some of the inputs, which are now documented.
- I've updated the infra template repository to match these changes, and rebased your rollout test branch.

Let me know if it works well

**Comment by djbe** [comment]
@daanpersoons-appwise As discussed, removed the fixed naming scheme from the custom action, it now just uses [the folder name as a bucket name](https://github.com/wisemen-digital/devops-ga-s3-apply/blob/main/action.yml#L47). I've also updated your [test branch](https://github.com/wisemen-digital/infra-project-template/tree/feature/s3-rollout) on the infra template repo.

**Review by djbe — APPROVED** [review]
